File: lang/ruby/bench_time.rb
#!/usr/bin/env ruby
$LOAD_PATH.unshift File.join(__dir__, 'cxlib', 'lib')
require 'cxlib'

medium = File.read(File.join(__dir__, '..', '..', 'fixtures', 'bench', 'bench_medium.cx'))

def time_median(n, warmup, &block)
  warmup.times { block.call }
  times = Array.new(n) do
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    block.call
    (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000
  end
  times.sort[n / 2]
end

parse_med  = time_median(100, 20) { CXLib.parse(medium) }
stream_med = time_median(100, 20) { CXLib.stream(medium) }
puts "parse=#{parse_med.round(3)} stream=#{stream_med.round(3)}"

File: lang/typescript/bench_time.ts
#!/usr/bin/env tsx
import * as fs from 'fs';
import * as path from 'path';
import { parse } from './cxlib/src/ast';
import { stream } from './cxlib/src/index';

const medium = fs.readFileSync(
    path.join(__dirname, '..', '..', 'fixtures', 'bench', 'bench_medium.cx'), 'utf-8');

function timeMedian(n: number, warmup: number, fn: () => void): number {
    for (let i = 0; i < warmup; i++) fn();
    const times: number[] = [];
    for (let i = 0; i < n; i++) {
        const t0 = performance.now();
        fn();
        times.push(performance.now() - t0);
    }
    times.sort((a, b) => a - b);
    return times[Math.floor(n / 2)];
}

const parseMed  = timeMedian(100, 20, () => { parse(medium); });
const streamMed = timeMedian(100, 20, () => { stream(medium); });
process.stdout.write(`parse=${parseMed.toFixed(3)} stream=${streamMed.toFixed(3)}\n`);
